### PR TITLE
DEPS: Pinning versions of some dependencies to speed up environment creation

### LINKED ISSUE #24012: Development conda environment taking >45 minutes to solve
When creating a new environment from our `environment.yml`, it takes more than 45 minutes to solve (one of the steps of the environment creation). This happens both locally, and in the CI of #22854, which uses that file for one of the builds.

```
2018-11-30T03:02:59.3313236Z [create env]
2018-11-30T03:49:30.3527900Z Solving environment: ...working... done
```

The number of dependencies in this environment is slightly higher than in the travis jobs (it has the packages for linting and building the docs), but there are not many more.

The environment with those dependencies was creating in a reasonable time until recently (like one week ago). We just added to it couple of small changes, that if undone, don't solve the problem. I also tried to downgrade conda, to see if a new release was doing something different, but the creation of the environment also takes the same with old versions.

I did a bit of research inside conda, and seems like the time is spent in `Resolve.solve`, in the step of maximizing the timestamps: https://github.com/conda/conda/blob/master/conda/resolve.py#L1031

This receives all the packages candidates as a parameter if I'm not wrong, so I assume the problem comes from a release in one of our dependencies, that makes this optimization very slow. I tried to remove packages from the dependencies file to see if there is a single package that when removed fixes the problem, but after many tests I couldn't identify one, I think it's a combination of packages that causes the issue.

When running the process, one of the cores stay at 100%, confirming (I think) that it's the optimization itself what is taking long (and not a network access or other IO).

@TomAugspurger do you know what the problem can be? And what's the best way to report this to conda? I guess an issue in the conda GitHub, but not sure if there is a better way.

### PR DESCRIPTION
**Pull request by datapythonista**
- [X] closes #24012
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Conda is taking 45 minutes to create an environment with our `environment.yml`. While this is a bug in conda, it won't be fixed until the next version. So, a workaround for now is to pin the verion of some of the packages.

I'm also removing couple of packages that I don't think are useful in a local environment, or in the CI build that will use the file (for the linting and to build the docs).

### CODE REVIEW

**Comment by msarahan** [comment]
nice!  I'm glad this works.  I'll use this as an example for others until we get conda sorted out.

**Comment by TomAugspurger** [comment]
Thanks for tracking this down @datapythonista. I'm perfectly happy with pinning and removing some optional deps for the dev env.

Feel free to merge once you've verified that the solver is performing well on the CI.

**Comment by h-vetinari** [comment]
Wouldn't it work to just use `>=` (sufficiently high) instead of `=` to reduce the number of versions? That way the environments can stay current, while not having to deal with the combinatorial blow-up that's slowing down current conda.

**Comment by datapythonista** [comment]
@TomAugspurger the CI is not using `environment.yml` yet, it will after #22854. Not sure if it'd be better to make the fixes in that PR, so we can see that the CI works. But the changes are somehow unrelated. What do you think?

@h-vetinari I guess that should work to, and will make newer versions installed (and I assume by the time we could have too many dependencies again, conda will be fixes). I'll give it a try and see.

**Comment by datapythonista** [comment]
@jreback you can see the "Checks and docs" azure job timing out (or taking a lot of time to complete) here: #22854

Also, if you try to recreate your local environment using `environment.yml` you should experience the same (in Linux, in Mac seems like it worked for Tom)

**Comment by TomAugspurger** [comment]
And keep in mind that this is just for the dev environment.yml, not for any
of the CI jobs.

File: environment.yml [hunk 1]
```diff
@@ -4,22 +4,21 @@ channels:
...
 
   # development
-  - Cython>=0.28.2
+  - cython>=0.28.2
   - flake8
   - flake8-comprehensions
   - flake8-rst=0.4.2
   - gitpython
-  - hypothesis>=3.58.0
+  - hypothesis=3.82
```
[the remark this answers is not in the document]
Comment (jreback) [reply environment.yml]: meaning, does it affect the solving speed

**Comment by TomAugspurger** [comment]
Thanks!

**Pull request merged**